File: addons/GuideCS/Inputs/GUIDEInputJoyAxis2D.cs
using Godot;

namespace Guide;

// [Tool] //not htere in original gd script
[GlobalClass]
public partial class GUIDEInputJoyAxis2D : GUIDEInputJoyBase {
	private JoyAxis _x = JoyAxis.LeftX;
	[Export]
	public JoyAxis X {
		get => _x;
		set {
			if (_x == value) return;
			_x = value;
			EmitChanged();
		}
	}

	private JoyAxis _y = JoyAxis.LeftY;
	[Export]
	public JoyAxis Y {
		get => _y;
		set {
			if (_y == value) return;
			_y = value;
			EmitChanged();
		}
	}

	public override void _BeginUsage() {
		_State.JoyAxisStateChanged += _Refresh;
		_Refresh();
	}

	public override void _EndUsage() {
		_State.JoyAxisStateChanged -= _Refresh;
	}

	private void _Refresh() {
		_Value = new Vector3(_State.GetJoyAxisValue(JoyIndex, X), _State.GetJoyAxisValue(JoyIndex, Y), 0);
	}

	public override bool IsSameAs(GUIDEInput other) {
		return other is GUIDEInputJoyAxis2D joyAxis &&
			   joyAxis.X == X &&
			   joyAxis.Y == Y &&
			   joyAxis.JoyIndex == JoyIndex;
	}

	public override string ToString() => $"(GUIDEInputJoyAxis2D: x={X}, y={Y}, joy_index={JoyIndex})";

	public override string _EditorName() => "Joy Axis 2D";

	public override string _EditorDescription() => "The input from two Joy axes. Usually from a stick.";

	public override GUIDEAction.GUIDEActionValueType _NativeValueType() => GUIDEAction.GUIDEActionValueType.Axis2D;
}

File: addons/guide-cpp/src/inputs/guide_input_joy_axis_2d.cpp
#include "guide_input_joy_axis_2d.h"
#include "guide_input_state.h"

using namespace godot;

void GUIDEInputJoyAxis2D::_bind_methods() {
    ClassDB::bind_method(D_METHOD("get_x_axis"), &GUIDEInputJoyAxis2D::get_x_axis);
    ClassDB::bind_method(D_METHOD("set_x_axis", "axis"), &GUIDEInputJoyAxis2D::set_x_axis);
    ADD_PROPERTY(PropertyInfo(Variant::INT, "x"), "set_x_axis", "get_x_axis");

    ClassDB::bind_method(D_METHOD("get_y_axis"), &GUIDEInputJoyAxis2D::get_y_axis);
    ClassDB::bind_method(D_METHOD("set_y_axis", "axis"), &GUIDEInputJoyAxis2D::set_y_axis);
    ADD_PROPERTY(PropertyInfo(Variant::INT, "y"), "set_y_axis", "get_y_axis");

    ClassDB::bind_method(D_METHOD("_refresh"), &GUIDEInputJoyAxis2D::_refresh);
}

GUIDEInputJoyAxis2D::GUIDEInputJoyAxis2D() {
}

GUIDEInputJoyAxis2D::~GUIDEInputJoyAxis2D() {
}

void GUIDEInputJoyAxis2D::_begin_usage() {
    if (_state) {
        _state->connect("joy_axis_state_changed", Callable(this, "_refresh"));
        _refresh();
    }
}

void GUIDEInputJoyAxis2D::_end_usage() {
    if (_state) {
        _state->disconnect("joy_axis_state_changed", Callable(this, "_refresh"));
    }
}

void GUIDEInputJoyAxis2D::_refresh() {
    if (_state) {
        _value.x = _state->get_joy_axis_value(joy_index, (JoyAxis)x);
        _value.y = _state->get_joy_axis_value(joy_index, (JoyAxis)y);
    }
}

bool GUIDEInputJoyAxis2D::is_same_as(const Ref<GUIDEInput> &other) const {
    Ref<GUIDEInputJoyAxis2D> o = other;
    if (o.is_null()) return false;
    return o->x == x && o->y == y && o->joy_index == joy_index;
}

String GUIDEInputJoyAxis2D::_editor_name() const {
    return "Joy Axis 2D";
}

String GUIDEInputJoyAxis2D::_editor_description() const {
    return "The input from two Joy axes. Usually from a stick.";
}

int GUIDEInputJoyAxis2D::_native_value_type() const {
    return 2; // AXIS_2D
}